examples/machine_learning/geneticalgorithm.cpp

```diff
@@ -0,0 +1,181 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <arrayfire.h>
+#include <cstdio>
+
+using namespace af;
+
+array update(const array& searchSpace, const array& samplex, const array& sampley, const int nsamples, const int n)
+{
+    array result(nsamples, 1, f32);
+    result = searchSpace(sampley*n + samplex);
+    return result;
+}
+
+array selectFittest(const array& samplez, const int nsamples) {
+    //pick top 50% fittest
+    array sorted = sort(samplez);
+    array indices = where(samplez >= sorted.row(nsamples / 2).scalar<float>());
+
+    if (indices.elements() > nsamples / 2) {
+        indices = indices.rows(indices.elements() - nsamples / 2, indices.elements() - 1);
+    }
+
+    return indices;
+}
+
+void reproduce(array& searchSpace, array& samplex, array& sampley, array& samplez, const int nsamples, const int n) {
+    //Get fittest parents
+    array selection = selectFittest(samplez, nsamples);
+    array parentsx = samplex(selection);
+    array parentsy = sampley(selection);
+
+    //Divide selection in two
+    array parentsx1 = parentsx.rows(0, parentsx.elements() / 2 - 1);
+    array parentsx2 = parentsx.rows(parentsx.elements() / 2, parentsx.elements() - 1);
+    array parentsy1 = parentsy.rows(0, parentsy.elements() / 2 - 1);
+    array parentsy2 = parentsy.rows(parentsy.elements() / 2, parentsy.elements() - 1);
+
+    //Get crossover points (at which bit to crossover) and construct bit masks from them
+    array crossover = randu(nsamples / 4, u32) % (log((double)n) / log(2.0));
+    array lowermask = pow2(crossover.as(f32)).as(u32) - 1;
+    array uppermask = INT_MAX - lowermask;
+
+    //Create children as the cross between two parents
+    array children_x1 = (parentsx1 & uppermask) + (parentsx2 & lowermask);
+    array children_y1 = (parentsy1 & uppermask) + (parentsy2 & lowermask);
+
+    array children_x2 = (parentsx2 & uppermask) + (parentsx1 & lowermask);
+    array children_y2 = (parentsy2 & uppermask) + (parentsy1 & lowermask);
+
+    //Join two new sets
+    samplex = join(0, children_x1, children_x2);
+    sampley = join(0, children_y1, children_y2);
+
+    //Create mutant children
+    array mutantx = samplex.copy();
+    array mutanty = sampley.copy();
+
+    //Flip a random bit to vary the gene pool
+    int bits = (int)(log((double)n) / log(2.0));
```

Comment:
Use log2, move this to the beginning and use `bits` in the other place I mentioned log2.